src/utils.js

```diff
@@ -18,12 +20,30 @@ function createUndiciRequestOptions (request, signal) {
     path: request.url.pathname + request.url.search,
     method: request.method,
     body: request.body,
-    headers: request.headers,
+    headers: request.headers[kHeaders],
     signal
   }
 }
 
+// Sorts 1-dimensional headers array by traversing it, from the end, and
+// comparing each pair to the previous pair. If the `>` comparison for
+// the previous pair is true, remove previous pair, move it to the end,
+// and reset the counter so it can restart the traversal and not miss anything
+function sort1d (arr) {
+  let i = arr.length
+  while (i > 0) {
+    if (arr[i - 2] > arr[i]) {
+      arr.push(...arr.splice(i - 2, 2))
```

Comment:
this is extremely slow :/.

Reply:
Yeah but it's faster than the first attempt. I'm open to other ideas how to sort this list quicker; maybe I'll give it another shot myself today

Reply:
Binary search on insert/push (set/append) would probably be fastest.

[this guy](https://github.com/substack/node-sorted) or a subset of it would probably work great

Reply:
I wrote a in-place mergeSort implementation and its still not faster 😅

```js
function merge(arr, start, mid, end) {
  // console.log(`m(${start}, ${mid}, ${end}) -> ${arr.slice(start, end)}`)

  for (let i = mid; i < end; i+=2) {
    let p = i
    for (let j = i - 2; j >= start; j-=2) {
      if (arr[p] < arr[j]) {
        [
          arr[p], arr[p+1],
          arr[j], arr[j+1]
        ] = [
          arr[j], arr[j+1],
          arr[p], arr[p+1]
        ]
        p = j
      }
    }
  }

  // console.log(arr)
}


function mergeSort(arr, start, end) {
  // console.log(`mS(${start}, ${end})`)
  if (start !== end - 2) {
    let div = (start + end) / 2
    let mid = div % 2 === 0 ? div : div + 1
    mergeSort(arr, start, mid)
    mergeSort(arr, mid, end)
    merge(arr, start, mid, end)
  }
}
```
I think v8 has really optimized the spread operator and the `slice` operation. I'll stash this mergeSort in a gist for reference but unless someone can write a faster sort algorithm i'm happy shipping as is

Reply:
> Binary search on insert/push (set/append) would probably be fastest.
> 
> 
> 
> [this module](https://github.com/substack/node-sorted) or a subset of it would probably work great
> 
> 
> 
> or this, is less OOy.. https://github.com/mafintosh/sorted-array-functions

Good idea; this was recommended previously, but there is concern it will decrease append/set performance. I'm going to give it a shot next.

My intuition tells me this will be the best solution because any sort of append/set needs to iterate the list anyways to see if that header already exists; so sorting on insert might actually improve perf since the scan can stop once it reaches the spot the given header name needs to be inserted

Reply:
latest commit implements sort on insert!

Reply:
Nice!